tosfs/core.py

```diff
@@ -203,6 +203,55 @@ def info(
 
         return self._try_dir_info(bucket, key, path, fullpath)
 
+    def rmdir(self, path: str) -> None:
+        """Remove a directory if it is empty.
+
+        Parameters
+        ----------
+        path : str
+            The path of the directory to remove. The path should be in the format
+            `tos://bucket/path/to/directory`.
+
+        Raises
+        ------
+        FileNotFoundError
+            If the directory does not exist.
+        NotADirectoryError
+            If the path is not a directory.
+        TosfsError
+            If the directory is not empty,
+             or the path is a bucket.
+
+        Examples
+        --------
+        >>> fs = TosFileSystem()
+        >>> fs.rmdir("tos://mybucket/mydir/")
+
+        """
+        path = self._strip_protocol(path).rstrip("/") + "/"
+        bucket, key, _ = self._split_path(path)
+        if not key:
+            raise TosfsError("Cannot remove a bucket using rmdir api.")
+
+        if not self.exists(path):
+            raise FileNotFoundError(f"Directory {path} not found.")
+
+        if not self.isdir(path):
+            raise NotADirectoryError(f"{path} is not a directory.")
+
+        if len(self.ls(path, refresh=True, detail=False)) > 0:
```

Comment:
The exists and isdir will call one info ( which means object HEAD)  representively ? And the `self.ls` will call an extra OBJECT LIST, which is quite time consuming ?

Reply:
>> The exists and isdir will call one info ( which means object HEAD) representively ?

Yes, for now.

Some of the context within this is: There are some basic APIs here, such as the ones you mentioned - `exists`, `isdir`, `isfile`, `info`, `ls` and so on. There are dependencies among them in the default implementation of fsspec. To sum up: We can only start implementing and overriding from the most core methods. But some methods are available by default (although there are performance bottlenecks, the default invocation here is mainly to complete the logic of the current API. This is why currently we haven't given high priority to considering the issues of stability and performance.  ), as you can see. Some methods have performance issues when using the default implementation. These will also be in the gradual replacement list, such as the already created exists(#18) and #14.

Reply:
> And the self.ls will call an extra OBJECT LIST, which is quite time consuming ?

Actually, the current `ls` supports the `dir cache`, and it is controlled by the `refresh` parameter. The default value of this parameter is: `False`. Here, since deletion is involved and it is necessary to determine that the current `dir` is empty, therefore, the call to `ls` here will forcibly initiate a request to the `tos server`. Regarding this point, do we have a better option?

Reply:
Maybe list objects and limit 1 is a better solution?